README.md

```diff
@@ -18,189 +18,259 @@ _TODO before entering Phase 2._
+### Detailed design discussion
 
-We consider this a good approach for the first implementation phase and aim to switch to a multiple threads per Wasm Instance once the `shared` attributes are added to the Wasm spec. Sharing the same instance right now is blocked on that attribute. More data on it can be found in this paper: [Weakening WebAssembly](https://www.researchgate.net/publication/336447205_Weakening_WebAssembly).
+Threads are tricky to implement. This proposal relies on a specific WebAssembly
+convention in order to work correctly. Upon a call to `thread_spawn`, the WASI
+host must:
+1. instantiate the module again &mdash; this child instance will be used for the
+   new thread
+2. in the child instance, ensure that any `shared` memories are the same ones as
+   those of the parent
+3. in the child instance, import all of the same WebAssembly objects as the
+   parent
+4. optionally, spawn a new host-level thread (other spawning mechanisms are
+   possible)
+5. calculate a non-duplicate thread ID, `tid`
+6. in the new thread, call the child instance's start function with the thread ID
+   and the start argument: `start_func(tid, start_arg)`
+
+A WASI host that implements the above should be able to spawn threads for a
+variety of languages.
+
+#### Design choice: pthreads
+
+One of the goals of this API is to be able to support `pthreads` for C compiled
+to WebAssembly. Given a WASI host that implements `thread_spawn` as described
+above, what responsibility would the C language have (i.e., `libc`) to properly
+implement `pthreads`?
+
+`pthread_create` must not only call WASI's `thread_spawn` but is also
+responsible for setting up the new thread's stack, TLS/TSD space, and updating
+the `pthread_t` structure. This could be implemented by (ignoring error
+conditions):
+1. configure a `struct start_args` with the user's `void *(*start_func)(void
+   *)` and `void *start_arg` (as done natively) but also with `pthread_t *thread`
+2. call `malloc` (instead of `mmap`) to allocate TLS/TSD in the shared
+   WebAssembly memory
+3. having defined a static `wasilibc_thread_start` function that takes as
+   parameters `int tid` and `void *start_args` , then in `pthread_create`, call
+   the WASI `thread_spawn` with this function and the configured `start_args`
+4. use `atomic.wait` to wait for the `start_args->thread->tid` value to change
+5. now in the child thread: once the WASI host creates the new thread instance
+   and calls `wasilibc_thread_start`, then a) set `args->thread->tid` to the
+   host-provided `tid`, b) set the `__wasilibc_pthread_self` global to point to
+   `args->thread` (this is used by `pthread_self`, e.g.), c) use `atomic.notify`
```

Comment:
IIRC this is not a problem since atomic.wait includes a value to compare against.  It will only wait if the value matches what is passsed in.  If that value has already been updated by the child thread it won't wait at all: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/wait